scalableminds/webknossos
# Iterate on version history view

This PR refines the version view as described in #3309. Versions are grouped by day and collapsed into 5-minute buckets. If a 5-min bucket would only contain a single version, it will be displayed directly.
A version 0 (`createTracing`) is added to the version view, so it is possible to revert to the very first state of the tracing.
This PR also minimizes (to some extent) the number of update actions that are being sent to the server. We're no longer unconditionally sending `updateTracing` update actions, if nothing changed. Additionally, subsequent update action batches that only contain an `updateTree` update action for the same treeId (happens for every keystroke when editing a comment, for example) will be "compressed", so that only the last `updateTree` update action is sent to the server.

### Open issues:
Ideally, I would like to jump to the respective volume tracing's `editPosition` when previewing previous volume tracing versions **in a hybrid tracing**, so the user can see the changed buckets, but I can only think of hacky ways to do that (by default only the skeleton tracing editPosition is used, and the volume one is discarded). Maybe you have a good idea for that?

When previewing volume versions (very) fast, bucket request errors are printed to the console (caused by the pending requests), however, I could not see any display artifacts. I also made sure that previewing volume versions doesn't cause a memory leak.

### Screenshots:
![refined-versions](https://user-images.githubusercontent.com/1702075/47017448-914bfd80-d152-11e8-8294-e56133f397a6.png)

![create-tracing](https://user-images.githubusercontent.com/1702075/47017552-d708c600-d152-11e8-9bfa-0a29fe4f4b9f.png)

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Open a hybrid tracing with some versions, or create one and skeleton/volume trace a bit. Then return 10 minutes later and trace some more. Then, opening the Version View should show the created versions in the previously described manner.
- Open an older skeleton tracing and open the version view. You should see the versions grouped by day.
- Edit a comment and then open the version view, only one version should have been created.

### Issues:
- fixes #3309 

------
